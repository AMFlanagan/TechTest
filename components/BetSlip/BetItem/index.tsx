import React from 'react'
import Icon from '../../common/Icon'
import { IconName } from '../../../constants/icons'
import {
  ActionTypes,
  IBetSlipEntry,
  useBetSlip,
} from '../../../context/BetSlip'
import placeBet from '../../../utils/placeBet'

import styles from './index.module.scss'
import formatNumbers from '../../../utils/formatNumbers'

interface IProps {
  bet: IBetSlipEntry
}

const BetItem: React.FC<IProps> = (props: IProps) => {
  const { dispatch } = useBetSlip()

  const [stake, setStake] = React.useState('')
  const [odds, setOdds] = React.useState('')

  const clickHandler = () => {
    // Would be call to route with validity check
    const res = placeBet({
      ...props.bet,
      stake: formatNumbers(stake),
      odds: formatNumbers(odds),
    })
    if (res) {
      removeBet()
    }
  }

  const handleChange = (key: string, value: string) => {
    if (key === 'stake') {
      setStake(value)
    }
    if (key === 'odds') {
      setOdds(value)
    }
  }

  const removeBet = () => {
    dispatch({
      type: ActionTypes.REMOVE,
      payload: {
        marketId: props.bet.marketId,
        outcome: props.bet.outcome,
      },
    })
  }

  return (
    <div className={styles.betItemContainer}>
      <div className={styles.top}>
        <div className={styles.info}>
          <div className={styles.outcomeName}>{props.bet.outcome}</div>
          <div className={styles.market}>MarketId: {props.bet.marketId}</div>
          <button onClick={() => removeBet()} className={styles.remove}>
            remove <Icon iconName={IconName.CROSS} />
          </button>
        </div>

        <div className={styles.actions}>
          <label>Stake</label>
          <input
            data-testid="stake"
            name="stake"
            value={stake}
            onChange={(e) => handleChange(e.target.name, e.target.value)}
            className={styles.stake}
            type="number"
            onBlur={() => setStake(formatNumbers(stake))}
          />
          <label>Odds</label>
          <input
            data-testid="odds"
            name="odds"
            value={odds}
            onChange={(e) => handleChange(e.target.name, e.target.value)}
            className={styles.odds}
            type="number"
            onBlur={() => setOdds(formatNumbers(odds))}
          />
        </div>
      </div>
      <button className={styles.placeBet} onClick={() => clickHandler()}>
        Place Bet
      </button>
    </div>
  )
}

BetItem.displayName = 'BetItem'
export default BetItem
