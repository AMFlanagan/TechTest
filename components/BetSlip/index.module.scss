.betSlipContainer {
  position: fixed;
  display: flex;
  flex-direction: column;
  background-color: #222;
  border: 2px solid #444;
  border-radius: 10px 10px 0 0;
  bottom: 0;
  margin-left: -18px;
  width: 100%;
  height: 100%;
  max-height: 500px;
  transition-duration: 0.3s;
}

.betSlipContainer > * {
  z-index: 1;
}

.betSlipContainer.closed {
  transform: translateY(450px);
  transition-duration: 0.3s;
}

.header {
  z-index: 9999;
  background: none;
  color: inherit;
  border: none;
  padding: 0;
  font: inherit;
  cursor: pointer;
  outline: inherit;

  height: 50px;
  width: 100%;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
}

.title,
.icon {
  margin: 16px;
}

.icon {
  transform: rotate(180deg);
  transition-duration: 0.3s;
}

.icon.closed {
  transform: rotate(0deg);
  transition-duration: 0.3s;
}

.base {
  height: 100px;
  margin: 0 auto;
  padding-top: 10px;
}

@media only screen and (min-width: 600px) {
  .betSlipContainer {
    max-width: 400px;
    right: 0;
    margin: 0;
  }
}
