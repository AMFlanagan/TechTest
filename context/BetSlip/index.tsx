import React from 'react'

export interface IBetSlipEntry {
  eventId: number
  eventName: string
  market: string
  marketId: number
  outcomeId: number
  outcomeName: string
  outcomeOdds: number
  stake: string
}

interface IState {
  bets: IBetSlipEntry[]
}

export enum ActionTypes {
  ADD = 'ADD',
  REMOVE = 'REMOVE',
  UPDATE = 'UPDATE',
  CLEAR_ALL = 'CLEAR_ALL',
}

type IAction = {
  type: ActionTypes
  payload?: any
}

const initialState = {
  bets: [],
}

interface IBetSlipContext {
  state: IState
  dispatch: (action: IAction) => IState | void
}

export const BetSlipContext = React.createContext<IBetSlipContext>({
  state: initialState,
  dispatch: null as any,
})

// TODO action type
export const bestSlipReducer = (state: IState, action: IAction): IState => {
  switch (action.type) {
    case ActionTypes.ADD: {
      return {
        ...state,
        bets: [...state.bets, action.payload],
      }
    }
    case ActionTypes.UPDATE: {
      const bets = [...state.bets]
      var foundIndex = bets.findIndex(
        (bet) =>
          bet.eventId === action.payload.eventId &&
          bet.outcomeId === action.payload.outcomeId
      )
      bets[foundIndex].stake = action.payload.stake
      return {
        ...state,
        bets,
      }
    }
    case ActionTypes.REMOVE: {
      const newBets = state.bets.filter(
        (bet) =>
          !(
            bet.eventId === action.payload.eventId &&
            bet.outcomeId === action.payload.outcomeId
          )
      )

      return {
        ...state,
        bets: newBets,
      }
    }
    case ActionTypes.CLEAR_ALL: {
      return {
        ...state,
        bets: [],
      }
    }
  }
}

export const BetSlipProvider = ({ children }: { children: any }) => {
  const [state, dispatch] = React.useReducer(bestSlipReducer, initialState)
  const value = { state, dispatch }

  return (
    <BetSlipContext.Provider value={value}>{children}</BetSlipContext.Provider>
  )
}

export const useBetSlip = () => {
  const context = React.useContext(BetSlipContext)
  if (context === undefined) {
    throw new Error('useBetSlip must be used within a BetSlipContext')
  }
  return context
}
