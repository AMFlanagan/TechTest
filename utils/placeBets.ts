export default (bets: any) => {
  // axios post request
  alert(`bets placed with ${JSON.stringify(bets)}`)
  return true
}
