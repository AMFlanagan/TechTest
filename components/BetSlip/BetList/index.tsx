import { IBetSlipEntry } from '../../../context/BetSlip'
import React from 'react'
import BetItem from '../BetItem'

import styles from './index.module.scss'

interface IProps {
  bets: IBetSlipEntry[]
}

const BetList: React.FC<IProps> = (props: IProps) => {
  return (
    <div className={styles.betListContainer}>
      {props.bets.length === 0 ? (
        <div className={styles.noBet}>Nothing in Betslip</div>
      ) : (
        props.bets.map((bet: IBetSlipEntry, i: number) => (
          // This wouldn't be an issue if all odds and stake state was
          // handled in the context as was planned
          <BetItem key={`${bet.marketId}${bet.outcome}`} bet={bet} />
        ))
      )}
    </div>
  )
}

BetList.displayName = 'BetList'
export default BetList
