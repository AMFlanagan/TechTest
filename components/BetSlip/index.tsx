import React from 'react'

import BetList from './BetList'
import { IconName } from '../../constants/icons'
import Icon from '../common/Icon'

import { useBetSlip } from '../../context/BetSlip'

import styles from './index.module.scss'

enum ViewState {
  OPEN = 'OPEN',
  CLOSED = 'CLOSED',
}

const BetSlip: React.FC = () => {
  const { state } = useBetSlip()
  const [viewState, setViewState] = React.useState(ViewState.CLOSED)

  return (
    <div
      className={`${styles.betSlipContainer} ${
        viewState === ViewState.CLOSED && styles.closed
      }`}
    >
      <button
        className={styles.header}
        onClick={() =>
          viewState === ViewState.OPEN
            ? setViewState(ViewState.CLOSED)
            : setViewState(ViewState.OPEN)
        }
      >
        <div className={styles.title}>Bet Slip</div>
        <div
          className={`${styles.icon} ${
            viewState === ViewState.CLOSED && styles.closed
          }`}
        >
          <Icon iconName={IconName.UP} />
        </div>
      </button>
      <BetList bets={state.bets} />
    </div>
  )
}

BetSlip.displayName = 'BetSlip'
export default BetSlip
