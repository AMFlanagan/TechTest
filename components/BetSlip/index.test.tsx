import { fireEvent, render } from '@testing-library/react'
import placeBets from '../../utils/placeBets'

import BetSlip from '.'

const dispatchMock = jest.fn()

const mockHook = {
  state: {
    bets: [
      {
        eventId: 1,
        eventName: 'eventName',
        market: 'market',
        marketId: 'marketId',
        outcomeId: 1,
        outcomeName: 'outcomeName',
        outcomeOdds: 1,
        stake: '0',
      },
    ],
  },
  dispatch: dispatchMock,
}

jest.mock('../../context/BetSlip', () => ({
  useBetSlip() {
    return mockHook
  },
}))

jest.mock('./BetList', () => jest.fn(() => <div>BetList</div>))
jest.mock('../../utils/placeBets')

describe('BetSlip', () => {
  it('renders as expected', () => {
    const { asFragment, getByText } = render(<BetSlip />)

    expect(getByText('BetList')).toBeTruthy()
    expect(asFragment()).toMatchSnapshot()
  })

  it('clicking header opens betslip', () => {
    const { asFragment, getByText } = render(<BetSlip />)

    const firstSnapShot = asFragment()
    fireEvent.click(getByText('Bet Slip'))
    const secondSnapShot = asFragment()

    expect(firstSnapShot).not.toEqual(secondSnapShot)
  })

  it('clicking header opens betslip', () => {
    const { getByText } = render(<BetSlip />)

    fireEvent.click(getByText('Place Bets'))

    expect(placeBets).toBeCalledWith([
      {
        eventId: 1,
        eventName: 'eventName',
        market: 'market',
        marketId: 'marketId',
        outcomeId: 1,
        outcomeName: 'outcomeName',
        outcomeOdds: 1,
        stake: '0',
      },
    ])
  })
})
