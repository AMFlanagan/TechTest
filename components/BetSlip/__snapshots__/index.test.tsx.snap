// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BetSlip renders as expected 1`] = `
<DocumentFragment>
  <div
    class="betSlipContainer closed"
  >
    <button
      class="header"
    >
      <div
        class="title"
      >
        Bet Slip
      </div>
      <div
        class="icon closed"
      >
        <div
          class="icon"
        >
          <svg
            height="512"
            id="Outline"
            viewBox="0 0 24 24"
            width="512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M23.71,16.29,15.54,8.12a5,5,0,0,0-7.08,0L.29,16.29a1,1,0,0,0,1.42,1.42L9.88,9.54a3,3,0,0,1,4.24,0l8.17,8.17a1,1,0,0,0,1.42,0A1,1,0,0,0,23.71,16.29Z"
            />
          </svg>
        </div>
      </div>
    </button>
    <div>
      BetList
    </div>
  </div>
</DocumentFragment>
`;
